### ISSUE TITLE
Task Not Serializable - NotSerializableException: scalapb.descriptors.FieldDescriptor

### ISSUE THREAD

**Issue by cdecoux** [scalapb/sparksql-scalapb#278]
I'm running into an issue with [Databricks runtime 10.4 (Spark 3.2.1, Scala 2.12)](https://docs.databricks.com/release-notes/runtime/10.4.html). I'm wondering if it's a solvable environmental issue with Databricks or inherently due to their proprietary Spark runtime, but I'm not even sure where to begin looking at this point. My test files/notebooks are below:

---
test/test.proto:
```proto
syntax = "proto3";
package cdecoux.test;
option java_multiple_files = true;
option java_outer_classname = "TestProto";
option java_package = "cdecoux.test";

message TestMessage {
  string test_message = 1;
}
```
---
plugins.sbt:
```sbt
addSbtPlugin("com.eed3si9n" % "sbt-assembly" % "1.1.0")
addSbtPlugin("com.thesamet" % "sbt-protoc" % "1.0.6")

/*
 List of dependency versions for the below version of scalapb:
 https://github.com/scalapb/ScalaPB/blob/v0.11.11/project/Dependencies.scala
 */
libraryDependencies += "com.thesamet.scalapb" %% "compilerplugin" % "0.11.11"
```

build.sbt
```sbt
organization := "cdecoux"
name := "test"
version := "0.0.1-SNAPSHOT"
scalaVersion := "2.12.10"
ThisBuild / publishMavenStyle := true
val sparkVersion = "3.2.1"

PB.includePaths in Compile := Seq(
  (ThisBuild / baseDirectory).value / ".." / ".." / "test"
)

Compile / PB.protoSources := Seq(
  (ThisBuild / baseDirectory).value / ".." / ".." / "test"
)

Compile / PB.targets := Seq(
  scalapb.gen() -> (Compile / sourceManaged).value / "scalapb"
)

libraryDependencies ++= Seq(
  "org.apache.spark" %% "spark-core" % sparkVersion % "provided",
  "org.apache.spark" %% "spark-sql" % sparkVersion % "provided",
  "com.thesamet.scalapb" %% "scalapb-runtime" % scalapb.compiler.Version.scalapbVersion,
  "com.thesamet.scalapb" %% "scalapb-runtime-grpc" % scalapb.compiler.Version.scalapbVersion,
  "com.thesamet.scalapb" %% "sparksql32-scalapb0_11" % "1.0.0", // https://scalapb.github.io/docs/sparksql#setting-up-your-project
)

assemblyShadeRules in assembly := Seq(
  ShadeRule.rename("com.google.protobuf.**" -> "shadeproto.@1").inAll,
  ShadeRule.rename("scala.collection.compat.**" -> "scalacompat.@1").inAll,
  ShadeRule.rename("shapeless.**" -> "shadeshapeless.@1").inAll
)

excludeFilter in Compile := "index.html"

assemblyMergeStrategy in assembly := {
  case x if x.contains("io.netty.versions.properties") => MergeStrategy.discard
  case x =>
    val oldStrategy = (assemblyMergeStrategy in assembly).value
    oldStrategy(x)
}
```

---
Notebook 1 (Deserializing bytes):
```scala
import scalapb.spark.Implicits._
import scalapb.{GeneratedMessage, GeneratedMessageCompanion}
import org.apache.spark.sql.Dataset
import scalapb.spark.Implicits.typedEncoderToEncoder
import cdecoux.test.test.TestMessage
import scalapb.spark.ProtoSQL


val events: Seq[TestMessage] = Seq[TestMessage](TestMessage(
  testMessage = "abc"
))
val bytes = events.map(_.toByteArray)

val protoDS = spark.createDataset(bytes)
val protoDF = ProtoSQL.protoToDataFrame(spark, protoDS.map(TestMessage.parseFrom)(typedEncoderToEncoder[TestMessage]).rdd)(typedEncoderToEncoder[TestMessage])

protoDF.show()
```

Notebook 2 (Just going from RDD to DF/DS):
```scala
import org.apache.spark.rdd.RDD
import scalapb.spark.Implicits._
import scalapb.{GeneratedMessage, GeneratedMessageCompanion}
import org.apache.spark.sql.Dataset
import scalapb.spark.Implicits.typedEncoderToEncoder
import cdecoux.test.test.TestMessage
import scalapb.spark.ProtoSQL


val events: Seq[TestMessage] = Seq[TestMessage](
  TestMessage(
    testMessage = "abc"
  ),
  TestMessage(
    testMessage = "zyx"
  ),
  TestMessage(
    testMessage = "xyz"
  )
)
val protoRDD: RDD[TestMessage] = spark.sparkContext.parallelize(events)
val protoDF = ProtoSQL.protoToDataFrame(spark, protoRDD)(typedEncoderToEncoder[TestMessage])
val protoDS = spark.createDataset(protoRDD)(typedEncoderToEncoder[TestMessage])

// Both fail
protoDF.show()
protoDS.show()
```

---
This gives me the `Job aborted due to stage failure. Caused by: NotSerializableException: scalapb.descriptors.FieldDescriptor` error along with a huge serialization stack. 

I did the same test in a local spark-shell and did not get the same error.

This seems to work with just `ProtoSQL.createDataframe(...)`

```
val protoDF = ProtoSQL.createDataFrame(spark, events)
protoDF.repartition(3).show()

(2) Spark Jobs
protoDF:org.apache.spark.sql.DataFrame = [test_message: string]
+------------+
|test_message|
+------------+
|         zyx|
|         abc|
|         xyz|
+------------+
```

**Comment by thesamet** [comment]
Hi @cdecoux , can you check if the same works outside databricks? One way to do that would be to add the above code to `PersonSpec` or `DefaultSpec` within this project and run it from sbt.

**Comment by cdecoux** [comment]
@thesamet Yes, I confirmed it works outside of Databricks. I imported the fat jar in spark shell via `spark-shell --jars` and ran the notebook code exactly (which actually worked).

**Comment by thesamet** [comment]
Is there a version on spark/databricks where the same example stopped working? Would it be possible for you to work with databricks support to suggest workarounds?

**Comment by cdecoux** [comment]
I was able to confirm that moving from Databricks Runtime LTS version 9.1 to 10.4 is when the issue arises. I will indeed open a discussion with Databricks Support regarding it.

**Comment by cdecoux** [comment]
Databricks support team is reporting that a fix should be available by the end of the month.

**Comment by pra91** [comment]
Just FYI this is also broken (and subsequently fixed) for open source spark 3.3.0, Issue is this one https://issues.apache.org/jira/browse/SPARK-40380

**Comment by cdecoux** [comment]
Databricks has deployed the fix and 10.4 DBR now seems to be working.

**Comment by satyam-raj-groww** [comment]
We have hit this issue when running on Google Dataproc image 2.1.9-debian11
Any resolutions?

**Comment by satyam-raj-groww** [comment]
Dataproc latest version is 2.1.x which comes with Spark 3.3.0 (ref)[https://cloud.google.com/dataproc/docs/concepts/versioning/dataproc-release-2.1]

This issue is resolved with Spark 3.3.1 (I ran with both spark versions on my local) which unfortunately Dataproc still has not released.

**Comment by thesamet** [comment]
As the bug is on the spark side, we'd wait for dataproc to release an image that contains a later Spark version.
